### PR TITLE
expand use of pytest and add a script to compare data files

### LINKED ISSUE #11: create a script or process to compare two datasets
- we often run a build of a data product/dataset and want to know if anything changed
- this is more black and white than QAQC
- there's already a `compare_csvs.py` in this repo, but we often generate SQL-format dump files.

### PR DESCRIPTION
**Pull request by damonmcc**
resolves #11 

## changes
- add `compare_data_files.py` to compare two tabular datasets
- add `utils.load_data_file` to load csv and json data
- add `test_utils.py` and two test data files for it to use
- rename test file from `test_sanity.py` to  `test_exmaples.py` and expand it to include examples of:
  - custom test markers
  - pytest parameters
  - testing for raised errors
- remove hardcoded path from `dev_compile_python_packages.sh`
- update `trunk` configuration versions
- report python test coverage when running `pytest`
- add `pyproject.toml` to configure python tests

## notes
- test coverage of this repo is currently at 15% [[link to pytest logs](https://github.com/NYCPlanning/db-template-repo/actions/runs/4834674734/jobs/8616191772#step:5:47)]
- the approaches in `compare_data_files.py` have been used to compare pluto corrections files (csv) and EDDE metadata files (json)

### CODE REVIEW

File: .github/workflows/tests_branch.yml [hunk 1]
```diff
@@ -46,7 +46,7 @@ jobs:
 
       - name: Run python tests
         run: |
-          python3 -m pytest
+          python3 -m pytest --verbose --verbose --cov-config=pyproject.toml --cov .
```
Comment (fvankrieken) [inline .github/workflows/tests_branch.yml:49]: I'm getting you on little stuff today but here we go: double `--verbose`

File: python/compare_data_files.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+INDEX_COLUMN = "pff_variable"
+
+
+def compare_data(old_data: pd.DataFrame, new_data: pd.DataFrame) -> None:
+    print("PREVIEW OLD DATA")
+    print(old_data.head().to_markdown())
+    old_data.info()
+    print("PREVIEW NEW DATA")
+    new_data.info()
+    print(new_data.head().to_markdown())
```
Comment (fvankrieken) [inline python/compare_data_files.py:23]: I did not know about `to_markdown` for a df, thank you for bringing this knowledge to me

File: python/compare_data_files.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        common_indices = set(old_data_raw[INDEX_COLUMN].unique()).intersection(
+            set(new_data_raw[INDEX_COLUMN].unique())
+        )
+
+        print("compare common rows ...")
+        print(f"{len(common_indices)} common index values")
+        old_data_to_compare, new_data_to_compare = limit_rows_to_compare(
+            old_data_raw, new_data_raw, INDEX_COLUMN, common_indices
+        )
+        compare_data(old_data_to_compare, new_data_to_compare)
```
Comment (fvankrieken) [inline python/compare_data_files.py:157]: Just commenting on this file as a whole, I like it as a start to this stuff. I'd imagine there might be some decently mature tools/packages to do some of this too. This sort of stuff should definitely be a module in our future internally published package (if we go that route obvs)

File: python/utils.py [hunk 4]
```diff
@@ -7,9 +8,14 @@
...
-    data = pd.read_csv(f"{directory}/{filename}")
-    data.columns = data.columns.str.lower()
+def load_data_file(filepath: str) -> pd.DataFrame:
+    file_extension = Path(filepath).suffix
+    if file_extension == ".csv":
+        data = pd.read_csv(filepath)
+    elif file_extension == ".json":
+        data = pd.read_json(filepath)
+    else:
+        raise NotImplementedError(f"Unsopported data file extension: {file_extension}")
```
Comment (fvankrieken) [inline python/utils.py:18]: Minor note but just enjoy seeing specific error types - haven't done a lot of it with Python but as we build stuff out would love to get error handling more specific and helpful

File: .github/workflows/tests_branch.yml (hunk 1 above)
Comment (fvankrieken) [reply .github/workflows/tests_branch.yml:49]: Oh wow! That's great actually, want a more verbose mode? Just throw another verbose in there

**Review by fvankrieken — APPROVED** [review]

File: python/compare_data_files.py (hunk 3 above)
Comment (damonmcc) [reply python/compare_data_files.py:157]: no doubt, this is mainly an approach I landed on at my last job that relies on `DataFrame.compare`

**Pull request merged**